File: services/academic-service/src/test/java/com/visor/school/academicservice/repository/ClassRepositoryTest.java
package com.visor.school.academicservice.repository;

import com.visor.school.academicservice.model.ClassStatus;
import com.visor.school.academicservice.model.ClassType;
import com.visor.school.academicservice.model.Term;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
import org.springframework.test.context.ActiveProfiles;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

@DataJpaTest
@ActiveProfiles("test")
class ClassRepositoryTest {

    @Autowired
    private TestEntityManager entityManager;

    @Autowired
    private ClassRepository classRepository;

    private com.visor.school.academicservice.model.Class testHomeroomClass;
    private com.visor.school.academicservice.model.Class testSubjectClass;

    @BeforeEach
    void setup() {
        testHomeroomClass = new com.visor.school.academicservice.model.Class("Grade 3 Homeroom", ClassType.HOMEROOM, null, 3, UUID.randomUUID(), null, "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null, ClassStatus.SCHEDULED);
        entityManager.persistAndFlush(testHomeroomClass);

        testSubjectClass = new com.visor.school.academicservice.model.Class("Mathematics 101", ClassType.SUBJECT, "Mathematics", 10, null, null, "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null, ClassStatus.SCHEDULED);
        entityManager.persistAndFlush(testSubjectClass);
    }

    @Test
    void shouldFindClassByClassName() {
        List<com.visor.school.academicservice.model.Class> found = classRepository.findByClassName("Grade 3 Homeroom");
        assertFalse(found.isEmpty());
        assertTrue(found.stream().anyMatch(c -> "Grade 3 Homeroom".equals(c.getClassName())));
    }

    @Test
    void shouldFindClassesByGradeLevel() {
        List<com.visor.school.academicservice.model.Class> grade3Classes = classRepository.findByGradeLevel(3);
        List<com.visor.school.academicservice.model.Class> grade10Classes = classRepository.findByGradeLevel(10);

        assertFalse(grade3Classes.isEmpty());
        assertTrue(grade3Classes.stream().allMatch(c -> c.getGradeLevel() == 3));
        assertFalse(grade10Classes.isEmpty());
        assertTrue(grade10Classes.stream().allMatch(c -> c.getGradeLevel() == 10));
    }

    @Test
    void shouldFindClassesByClassType() {
        List<com.visor.school.academicservice.model.Class> homeroomClasses = classRepository.findByClassType(ClassType.HOMEROOM);
        List<com.visor.school.academicservice.model.Class> subjectClasses = classRepository.findByClassType(ClassType.SUBJECT);

        assertFalse(homeroomClasses.isEmpty());
        assertTrue(homeroomClasses.stream().allMatch(c -> c.getClassType() == ClassType.HOMEROOM));
        assertFalse(subjectClasses.isEmpty());
        assertTrue(subjectClasses.stream().allMatch(c -> c.getClassType() == ClassType.SUBJECT));
    }

    @Test
    void shouldFindClassesByAcademicYearAndTerm() {
        List<com.visor.school.academicservice.model.Class> classes = classRepository.findByAcademicYearAndTerm("2024-2025", Term.FIRST_TERM);
        assertFalse(classes.isEmpty());
        assertTrue(classes.stream().allMatch(c -> "2024-2025".equals(c.getAcademicYear()) && c.getTerm() == Term.FIRST_TERM));
    }

    @Test
    void shouldFindHomeroomClassesByTeacher() {
        UUID homeroomTeacherId = testHomeroomClass.getHomeroomTeacherId();
        List<com.visor.school.academicservice.model.Class> classes = classRepository.findByHomeroomTeacherId(homeroomTeacherId);
        assertFalse(classes.isEmpty());
        assertTrue(classes.stream().allMatch(c -> homeroomTeacherId.equals(c.getHomeroomTeacherId())));
    }

    @Test
    void shouldFindSubjectClassesByClassTeacher() {
        UUID classTeacherId = UUID.randomUUID();
        com.visor.school.academicservice.model.Class classWithTeacher = new com.visor.school.academicservice.model.Class("English 101", ClassType.SUBJECT, "English", 11, null, classTeacherId, "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null, ClassStatus.SCHEDULED);
        entityManager.persistAndFlush(classWithTeacher);

        List<com.visor.school.academicservice.model.Class> classes = classRepository.findByClassTeacherId(classTeacherId);
        assertFalse(classes.isEmpty());
        assertTrue(classes.stream().allMatch(c -> classTeacherId.equals(c.getClassTeacherId())));
    }

    @Test
    void shouldFindClassesByStatus() {
        com.visor.school.academicservice.model.Class inProgressClass = new com.visor.school.academicservice.model.Class("In Progress Class", ClassType.SUBJECT, "Science", 8, null, null, "2024-2025", Term.FIRST_TERM, null, null, LocalDate.of(2024, 9, 1), null, ClassStatus.IN_PROGRESS);
        entityManager.persistAndFlush(inProgressClass);

        List<com.visor.school.academicservice.model.Class> inProgressClasses = classRepository.findByStatus(ClassStatus.IN_PROGRESS);
        assertFalse(inProgressClasses.isEmpty());
        assertTrue(inProgressClasses.stream().allMatch(c -> c.getStatus() == ClassStatus.IN_PROGRESS));
    }
}

File: services/academic-service/src/test/kotlin/com/visor/school/academicservice/repository/ClassRepositoryTest.kt
package com.visor.school.academicservice.repository

import com.visor.school.academicservice.model.*
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager
import org.springframework.test.context.ActiveProfiles
import java.time.LocalDate
import java.util.UUID

@DataJpaTest
@ActiveProfiles("test")
class ClassRepositoryTest @Autowired constructor(
    private val entityManager: TestEntityManager,
    private val classRepository: ClassRepository
) {

    private lateinit var testHomeroomClass: Class
    private lateinit var testSubjectClass: Class

    @BeforeEach
    fun setup() {
        testHomeroomClass = Class(
            className = "Grade 3 Homeroom",
            classType = ClassType.HOMEROOM,
            subject = null,
            gradeLevel = 3,
            homeroomTeacherId = UUID.randomUUID(),
            classTeacherId = null,
            academicYear = "2024-2025",
            term = Term.FIRST_TERM,
            startDate = LocalDate.of(2024, 9, 1)
        )
        entityManager.persistAndFlush(testHomeroomClass)

        testSubjectClass = Class(
            className = "Mathematics 101",
            classType = ClassType.SUBJECT,
            subject = "Mathematics",
            gradeLevel = 10,
            homeroomTeacherId = null,
            classTeacherId = null,
            academicYear = "2024-2025",
            term = Term.FIRST_TERM,
            startDate = LocalDate.of(2024, 9, 1)
        )
        entityManager.persistAndFlush(testSubjectClass)
    }

    @Test
    fun `should find class by className`() {
        val found = classRepository.findByClassName("Grade 3 Homeroom")

        assertTrue(found.isNotEmpty())
        assertTrue(found.any { it.className == "Grade 3 Homeroom" })
    }

    @Test
    fun `should find classes by grade level`() {
        val grade3Classes = classRepository.findByGradeLevel(3)
        val grade10Classes = classRepository.findByGradeLevel(10)

        assertTrue(grade3Classes.isNotEmpty())
        assertTrue(grade3Classes.all { it.gradeLevel == 3 })
        assertTrue(grade10Classes.isNotEmpty())
        assertTrue(grade10Classes.all { it.gradeLevel == 10 })
    }

    @Test
    fun `should find classes by class type`() {
        val homeroomClasses = classRepository.findByClassType(ClassType.HOMEROOM)
        val subjectClasses = classRepository.findByClassType(ClassType.SUBJECT)

        assertTrue(homeroomClasses.isNotEmpty())
        assertTrue(homeroomClasses.all { it.classType == ClassType.HOMEROOM })
        assertTrue(subjectClasses.isNotEmpty())
        assertTrue(subjectClasses.all { it.classType == ClassType.SUBJECT })
    }

    @Test
    fun `should find classes by academic year and term`() {
        val classes = classRepository.findByAcademicYearAndTerm("2024-2025", Term.FIRST_TERM)

        assertTrue(classes.isNotEmpty())
        assertTrue(classes.all { it.academicYear == "2024-2025" && it.term == Term.FIRST_TERM })
    }

    @Test
    fun `should find homeroom classes by teacher`() {
        val homeroomTeacherId = testHomeroomClass.homeroomTeacherId!!
        val classes = classRepository.findByHomeroomTeacherId(homeroomTeacherId)

        assertTrue(classes.isNotEmpty())
        assertTrue(classes.all { it.homeroomTeacherId == homeroomTeacherId })
    }

    @Test
    fun `should find subject classes by class teacher`() {
        val classTeacherId = UUID.randomUUID()
        val classWithTeacher = Class(
            className = "English 101",
            classType = ClassType.SUBJECT,
            subject = "English",
            gradeLevel = 11,
            homeroomTeacherId = null,
            classTeacherId = classTeacherId,
            academicYear = "2024-2025",
            term = Term.FIRST_TERM,
            startDate = LocalDate.of(2024, 9, 1)
        )
        entityManager.persistAndFlush(classWithTeacher)

        val classes = classRepository.findByClassTeacherId(classTeacherId)

        assertTrue(classes.isNotEmpty())
        assertTrue(classes.all { it.classTeacherId == classTeacherId })
    }

    @Test
    fun `should find classes by status`() {
        val inProgressClass = Class(
            className = "In Progress Class",
            classType = ClassType.SUBJECT,
            subject = "Science",
            gradeLevel = 8,
            homeroomTeacherId = null,
            classTeacherId = null,
            academicYear = "2024-2025",
            term = Term.FIRST_TERM,
            startDate = LocalDate.of(2024, 9, 1),
            status = ClassStatus.IN_PROGRESS
        )
        entityManager.persistAndFlush(inProgressClass)

        val inProgressClasses = classRepository.findByStatus(ClassStatus.IN_PROGRESS)

        assertTrue(inProgressClasses.isNotEmpty())
        assertTrue(inProgressClasses.all { it.status == ClassStatus.IN_PROGRESS })
    }
}

